### PR TITLE
ARROW-11997: [Python] concat_tables crashes python interpreter

### PR DESCRIPTION
**Pull request by dianaclarke**
Before:

```
>>> import pyarrow
>>> pyarrow.concat_tables([None])
Bus error: 10
Python quit unexpectedly
```

After:

```
>>> import pyarrow
>>> pyarrow.concat_tables([None])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pyarrow/table.pxi", line 2228, in pyarrow.lib.concat_tables
    c_tables.push_back(table.sp_table)
AttributeError: 'NoneType' object has no attribute 'sp_table
```

### CODE REVIEW

**Comment by pitrou** [comment]
This is a recurring problem, so instead of testing for None manually we could let Cython do it for us:
```diff
diff --git a/python/pyarrow/lib.pyx b/python/pyarrow/lib.pyx
index eba0f5a47..aa6918b54 100644
--- a/python/pyarrow/lib.pyx
+++ b/python/pyarrow/lib.pyx
@@ -15,9 +15,10 @@
 # specific language governing permissions and limitations
 # under the License.
 
-# cython: profile=False
-# distutils: language = c++
+# cython: profile = False
 # cython: embedsignature = True
+# cython: nonecheck = True
+# distutils: language = c++
 
 import datetime
 import decimal as _pydecimal
diff --git a/python/pyarrow/table.pxi b/python/pyarrow/table.pxi
index 2d474ba76..3f1fc28ee 100644
--- a/python/pyarrow/table.pxi
+++ b/python/pyarrow/table.pxi
@@ -2225,8 +2225,6 @@ def concat_tables(tables, c_bool promote=False, MemoryPool memory_pool=None):
             CConcatenateTablesOptions.Defaults())
 
     for table in tables:
-        if table is None:
-            raise TypeError("Unable to concatenate table. Table is None.")
         c_tables.push_back(table.sp_table)
 
     with nogil:
```

Note that a few tests will need fixing.

**Comment by dianaclarke** [comment]
Oh, thanks a ton @pitrou!

I'm working on the additional failing tests now.

File: python/pyarrow/tests/test_table.py [hunk 1]
```diff
@@ -271,7 +271,9 @@ def ne(xarrs, yarrs):
     eq([a, c], [d])
     ne([c, a], [a, c])
 
-    assert not pa.chunked_array([], type=pa.int32()).equals(None)
+    # ARROW-4822
+    with pytest.raises(AttributeError):
+        pa.chunked_array([], type=pa.int32()).equals(None)
```
Comment (dianaclarke) [inline python/pyarrow/tests/test_table.py:276]: This is a behaviour change though.

Before:
```
>>> import pyarrow
>>> table = pyarrow.chunked_array([], type=pyarrow.int32())
>>> table is None
False
>>> table == None
False
```

After:

```
>>> import pyarrow
>>> table = pyarrow.chunked_array([], type=pyarrow.int32())
>>> table is None
False
>>> table == None
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pyarrow/table.pxi", line 187, in pyarrow.lib.ChunkedArray.__eq__
    return self.equals(other)
  File "pyarrow/table.pxi", line 212, in pyarrow.lib.ChunkedArray.equals
    CChunkedArray* other_arr = other.chunked_array
AttributeError: 'NoneType' object has no attribute 'chunked_array'
```

Comment (pitrou) [reply python/pyarrow/tests/test_table.py:276]: Right, `table == None` should return False.

Comment (dianaclarke) [reply python/pyarrow/tests/test_table.py]: Behaviour is back to normal now.

Comment (dianaclarke) [reply python/pyarrow/tests/test_table.py]: Oh, I just saw your edit: `table.equals(None) can still raise TypeError though.`

Are you okay with this? This is how it was before.

```
>>> import pyarrow
>>> table = pyarrow.chunked_array([], type=pyarrow.int32())
>>> table.equals(None)
False
```

**Review by pitrou — APPROVED** [review]
+1, thank you @dianaclarke

**Comment by pitrou** [comment]
Green CI run at https://github.com/dianaclarke/arrow/runs/2131383253

**Comment by pitrou** [comment]
@dianaclarke You may want to enable Travis-CI on your Arrow fork for faster CI results.
(in this case this doesn't matter, since Travis-CI doesn't run Python tests)

**Pull request closed without merging**